### PR TITLE
[SPARK-41346][CONNECT][PYTHON] Implement `asc` and `desc` functions

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, remove `asc` and `desc` from `Expression`;
2, add `asc_nulls_first`, `asc_nulls_last`, `desc_nulls_first`, `desc_nulls_last` in `Column`;
3, add these methods in `functions`;
4, add a new test file for functions

### Why are the changes needed?
for api coverage

### Does this PR introduce _any_ user-facing change?
yes, new APIs

### How was this patch tested?
added ut

### CODE REVIEW

File: python/pyspark/sql/connect/column.py [hunk 1]
```diff
@@ -701,11 +688,23 @@ def to_plan(self, session: "SparkConnectClient") -> proto.Expression:
     def alias(self, *alias: str, **kwargs: Any) -> "Column":
         return Column(self._expr.alias(*alias, **kwargs))
 
+    def asc(self) -> "Column":
+        return self.asc_nulls_first()
```
Comment (amaliujia) [inline python/pyspark/sql/connect/column.py:692]: oh I didn't know `asc` is `asc_nulls_first`. This is a good simplification.

File: python/pyspark/sql/connect/functions.py [hunk 2]
```diff
@@ -16,14 +16,247 @@
...
+    """
+    return _to_col(col).asc_nulls_first()
+
+
+def asc_nulls_last(col: "ColumnOrName") -> Column:
+    """
+    Returns a sort expression based on the ascending order of the given
+    column name, and null values appear after non-null values.
+
+    .. versionadded:: 2.4.0
```
Comment (amaliujia) [inline python/pyspark/sql/connect/functions.py:131]: nit: wrong `versionadded`. BTW do we need always add `versionadded  3.4.0` given the entire package is `versionadded`? Maybe only need a top level `versionadded`?

**Review by amaliujia — APPROVED** [review]
LGTM with a minor comment

File: python/pyspark/sql/connect/column.py [hunk 3]
```diff
@@ -84,7 +85,7 @@ def __init__(self) -> None:
     def to_plan(self, session: "SparkConnectClient") -> "proto.Expression":
         ...
 
-    def __str__(self) -> str:
+    def __repr__(self) -> str:
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/column.py:88]: a simple example for why we should use `__repr__`:

```

In [20]: class A:
    ...:     def __str__(self) -> str:
    ...:         return "123"
    ...:

In [21]: a = A()

In [22]: a
Out[22]: <__main__.A at 0x1034bef40>

In [23]: str(a)
Out[23]: '123'

In [24]: repr(a)
Out[24]: '<__main__.A object at 0x1034bef40>'


In [30]: class A:
    ...:     def __repr__(self) -> str:
    ...:         return "123"
    ...:

In [31]: a = A()

In [32]: a
Out[32]: 123

In [33]: str(a)
Out[33]: '123'

In [34]: repr(a)
Out[34]: '123'


In [35]: class A:
    ...:     def __str__(self) -> str:
    ...:         return "123"
    ...:     def __repr__(self) -> str:
    ...:         return "xyz"
    ...:

In [36]: a = A()

In [37]: a
Out[37]: xyz

In [38]: str(a)
Out[38]: '123'

In [39]: repr(a)
Out[39]: 'xyz'
```

File: python/pyspark/sql/connect/functions.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py:131]: nice. will update.  I not sure whether we can set a top level `versionadded`

File: python/pyspark/sql/connect/functions.py [hunk 4]
```diff
@@ -16,14 +16,247 @@
...
+from typing import Any, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from pyspark.sql.connect._typing import ColumnOrName
 
 # TODO(SPARK-40538) Add support for the missing PySpark functions.
 
 
+def _to_col(col: "ColumnOrName") -> Column:
+    return col if isinstance(col, Column) else column(col)
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:28]:
```suggestion
    return col if isinstance(col, Column) else Column(col)
```

File: python/pyspark/sql/connect/functions.py [hunk 5]
```diff
@@ -16,14 +16,247 @@
...
+
+
 def col(x: str) -> Column:
     return Column(ColumnReference(x))
 
 
+column = col
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:35]: why???

Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py:35]: `column` itself is a public function.

https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L208

**Review by grundprinzip — CHANGES_REQUESTED** [review]
Please remove the nonsensical alias for `col = colum = Column`.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/functions.py (hunk 4 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py:28]: after the refactoring, `Column` no longer takes a str as input

**Comment by zhengruifeng** [comment]
thanks for reviews, merged into master

File: python/pyspark/sql/tests/connect/test_connect_function.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+from pyspark.testing.sqlutils import have_pandas, SQLTestUtils
+
+from pyspark.sql import SparkSession, Row
+from pyspark.sql.types import StructType, StructField, StringType
+
+if have_pandas:
+    from pyspark.sql.connect.session import SparkSession as RemoteSparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.testing.connectutils import should_test_connect, connect_requirement_message
+from pyspark.testing.pandasutils import PandasOnSparkTestCase
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/tests/connect/test_connect_function.py:30]: Hi, @zhengruifeng . This should be inside `if have_pandas`. I made a PR to fix it.
- https://github.com/apache/spark/pull/38929

**Pull request closed without merging**